### PR TITLE
Add the ability to add healthcheck definition for containers

### PR DESCRIPTION
**Pull request by svagner**
Original issue: https://github.com/bazelbuild/rules_docker/issues/770

Adds follow options for container_image:
- healthcheck_test
- healthcheck_interval
- healthcheck_timeout
- healthcheck_start_period
- healthcheck_retries

### CODE REVIEW

**Comment by svagner** [comment]
@googlebot I signed it!

**Comment by svagner** [comment]
cc @xingao267 , @alex1545 as you were in the discussion in the original issue

File: container/go/pkg/compat/config.go [hunk 1]
```diff
@@ -42,6 +42,29 @@ const (
 	defaultTimestamp = "1970-01-01T00:00:00Z"
 )
 
+var (
+	validHealcheckTypes = map[string]func([]string) error{
```
Comment (zoidbergwill) [inline container/go/pkg/compat/config.go:46]: nit

```suggestion
	validHealthcheckTypes = map[string]func([]string) error{
```

File: container/go/pkg/compat/config.go [hunk 2]
```diff
@@ -42,6 +42,29 @@ const (
...
+		"NONE": func(args []string) error {
+			if len(args) == 1 {
+				return nil
+			}
+			return errors.New("NONE doesn't accept any extra args")
```
Comment (zoidbergwill) [inline container/go/pkg/compat/config.go:51]: I'm not super familiar with Go, but I think it's common practice to put the error case in the indented if statement and have the default be the positive best case path.

**Comment by zoidbergwill** [comment]
Hiding whitespace changes definitely makes this diff easier to review: https://github.com/bazelbuild/rules_docker/pull/1742/files?w=1

File: container/go/pkg/compat/config.go (hunk 1 above)
Comment (svagner) [reply container/go/pkg/compat/config.go]: Fixed. Thanks!

File: container/go/pkg/compat/config.go (hunk 2 above)
Comment (svagner) [reply container/go/pkg/compat/config.go]: Yeah. I don't have a strong opinion on that so I've moved the error case under a conditional branch as you offered

File: README.md [hunk 3]
```diff
@@ -2195,6 +2195,82 @@ container_image(name, base, data_path, directory, files, legacy_repository_namin
...
     </tr>
+    <tr>
+      <td><code>healthcheck_test</code></td>
+      <td>
+        <code>List of strings, optional</code>
+        <p><a href="https://docs.docker.com/engine/reference/builder/#healthcheck">
+        The test to perform to check that the container is healthy.</a></p>
+        <p>An empty list means to inherit the default.</p>
+        <p>Possible definition options are:</p>
+        <p><code>[]</code> : inherit healthcheck</p>
```
Comment (gravypod) [inline README.md:2206]: `None` also inherits the healthcheck. It may be nice to specify this for callers who write their own macros wrapping this.

**Comment by gravypod** [comment]
I don't see a reason why this couldn't be merged. The only concern I have is that we're overloading `healthcheck_test` to do multiple types of tests (`CMD`, `CMD-SHELL`, `NONE`, etc). Ideally I'd gravitate towards separating these out but I'm not sure that would make sense in this case due to this being a docker-specific rule. 

I'm leaning towards thinking this is fine to include but I'd like the opinion of another maintainer just as a sanity check.

**Comment by pcj** [comment]
I think it's cool that you put this together, but I'm a little wary of this as-is for a couple of reasons:

1. It's a bit telling that I've never heard of this feature before today.  As far as I can tell, it's not supported in kubernetes but was a feature added on by Docker Inc for swarm specifically.  I don't think this feature is going to get much mileage, and it comes at a time when we're trying to simplify this repo rather than add complexity.
2. Given a `Dockerfile` only has the `HEALTHCHECK` command, why does `rules_docker` now need 5 new options to say the same thing?  If there is no parsing library in `google/go-containerregistry` (upon which this repo depends) to parse that into the necessary config, then that library/function should probably there upstream before merging this.  I don't think we should need more than a single new `string` or `string_list` attribute `healthcheck` for this.

File: README.md (hunk 3 above)
Comment (svagner) [reply README.md:2206]: Added. Thanks!

**Comment by svagner** [comment]
> I don't see a reason why this couldn't be merged. The only concern I have is that we're overloading `healthcheck_test` to do multiple types of tests (`CMD`, `CMD-SHELL`, `NONE`, etc). Ideally I'd gravitate towards separating these out but I'm not sure that would make sense in this case due to this being a docker-specific rule.
> 
> I'm leaning towards thinking this is fine to include but I'd like the opinion of another maintainer just as a sanity check.

Yeah. I also thought that probably it's better to have `healthcheck_cmd` and `healthcheck_cmd_shell` options. But it will lead to more complications:(

**Comment by svagner** [comment]
> I think it's cool that you put this together, but I'm a little wary of this as-is for a couple of reasons:
> 
> 1. It's a bit telling that I've never heard of this feature before today.  As far as I can tell, it's not supported in kubernetes but was a feature added on by Docker Inc for swarm specifically.  I don't think this feature is going to get much mileage, and it comes at a time when we're trying to simplify this repo rather than add complexity.
> 2. Given a `Dockerfile` only has the `HEALTHCHECK` command, why does `rules_docker` now need 5 new options to say the same thing?  If there is no parsing library in `google/go-containerregistry` (upon which this repo depends) to parse that into the necessary config, then that library/function should probably there upstream before merging this.  I don't think we should need more than a single new `string` or `string_list` attribute `healthcheck` for this.

Thanks for the feedback. Let me explain my thoughts about it.

1. These rules are called `rules_docker` so I believe it means that we should support at least all the main features of docker. I agree that kube doesn't use docker's `HEALTHCHECK` since it has it's own` livenessProbe` but this plugin isn't dedicated to kube only is it?
2. I think having `healthcheck_test =" --interval = 5s --timeout = 10s --retries = 3 CMD curl -sS 127.0.0.1 "" will be a little bit confusing for users of these rules since interval / to / retries are optional parameters for health checking right? And unfortunately I don't follow your point about the parsing of this parameter in the `google / go-containerregistry`. Since it is a single directive there is not necessary to parse it. But I'm again worrying about moving complications of the definition for health checks from the rules code to users.

Let me know if my arguments make sense to you.

**Comment by pcj** [comment]
I don't see how `--interval = 5s --timeout = 10s --retries = 3 CMD curl -sS 127.0.0.1` could possibly be confusing to any user wanting to use this, because that's the syntax Docker came up with and has clearly documented.

And it is definitely necessary to parse it, because the configuration for it is not a simple string.  It is https://github.com/moby/moby/blob/46cdcd206c56172b95ba5c77b827a722dab426c5/api/types/container/config.go#L16-L35:

```go
// HealthConfig holds configuration settings for the HEALTHCHECK feature.
type HealthConfig struct {
	// Test is the test to perform to check that the container is healthy.
	// An empty slice means to inherit the default.
	// The options are:
	// {} : inherit healthcheck
	// {"NONE"} : disable healthcheck
	// {"CMD", args...} : exec arguments directly
	// {"CMD-SHELL", command} : run command with system's default shell
	Test []string `json:",omitempty"`

	// Zero means to inherit. Durations are expressed as integer nanoseconds.
	Interval    time.Duration `json:",omitempty"` // Interval is the time to wait between checks.
	Timeout     time.Duration `json:",omitempty"` // Timeout is the time to wait before considering the check to have hung.
	StartPeriod time.Duration `json:",omitempty"` // The start period for the container to initialize before the retries starts to count down.

	// Retries is the number of consecutive failures needed to consider a container as unhealthy.
	// Zero means inherit.
	Retries int `json:",omitempty"`
}
```

Part of the parser is here: https://github.com/moby/buildkit/blob/master/frontend/dockerfile/parser/line_parsers.go#L343-L369

But rules_docker has no dependency on moby library, it has a dependency on go-containerregistry, which has the same copy of the config: https://github.com/moby/buildkit/blob/master/frontend/dockerfile/parser/line_parsers.go#L343-L369.  

Yes, as far as I know, there is no parsing logic in go-containerregistry to transform a string into that config.  The parsing logic should be upstreamed there, in the same place that has the config.

**Comment by svagner** [comment]
Ah, you meant about an unmarshaling line to the struct.  Yeah. Makes sense.  I'm not sure `go-containerregistry` will approve this change though since it looks like they don't need such a directive there. I can try to prepare PR for `go-containerregistry` for it but are we ready to add a health check directive here? I really just don't want to waste reviewers of  `go-containerregistry` time and my time if we don't want this feature at all.

> Also, rules_docker should not strive to replicate every feature in docker or a Dockerfile. A noteable absence is the RUN feature.

**Comment by pcj** [comment]
If we can get a parsing function upstreamed and simplify this PR to add only a single `healthcheck` attribute, I would support it.

**Comment by medzin** [comment]
@svagner will you continue this PR?

**Comment by alexeagle** [comment]
Hey @jonjohnsonjr what do you think about the proposed upstream change to go-containerregistry?

**Comment by jonjohnsonjr** [comment]
> what do you think about the proposed upstream change to go-containerregistry?

That sounds reasonable to me, something like `func ParseHealthConfig(s string) (*HealthConfig, error)`?

I'm happy to review a PR.

**Comment by alexeagle** [comment]
@dgocoder I'm not sure if @svagner is still around to finish this work (by upstreaming the Go parts) or if it needs a new contributor to step up to finish.

**Comment by nadavwe** [comment]
hey @pjjw, any luck?
We would really love to see this merged.
currently we need to have a base image created in a different build to include the healthcheck, quite far from the application definition... :/

**Comment by pjjw** [comment]
no, have not plumbed this upstream yet, but this diff works for us as-is.

**Pull request closed without merging**